### PR TITLE
Add pnum

### PR DESCRIPTION
**Pull request by JasonMHasperhoven**
Currently we have a bunch of conversion functions scattered around in:
- packages/types/src/amount.ts
- packages/types/src/lo-hi.ts
- packages/types/src/round.ts
- packages/types/src/value-view.ts

This aims to unify and simplify number conversions, see the function description at the top for context.

### CODE REVIEW

File: packages/types/src/pnum.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+ * - in base units for bigint, LoHi, Amount and ValueView
+ * - in display form with decimals for string and number
+ *
+ * @param input
+ * @param exponent
+ */
+function pnum(
+  input: string | number | LoHi | bigint | Amount | ValueView | undefined,
+  exponentInput = 0,
+) {
```
Comment (grod220) [inline packages/types/src/pnum.ts:29]: issue: this feels very class-like (private vars, exposed methods, constructor like initialization). Think we should convert it.

File: packages/types/src/pnum.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  } else if (typeof input === 'bigint') {
+    value = new BigNumber(input.toString());
+  } else if (input instanceof ValueView) {
+    const amount = getAmount(input);
+    value = new BigNumber(joinLoHi(amount.lo, amount.hi).toString());
+    exponent =
+      input.valueView.case === 'knownAssetId' ? getDisplayDenomExponentFromValueView(input) : 0;
+  } else if (
+    input instanceof Amount ||
+    (typeof input === 'object' && 'lo' in input && 'hi' in input && input.lo && input.hi)
```
Comment (grod220) [inline packages/types/src/pnum.ts:44]: Suggestion: Wouldn't this work the same?

```ts
else if (input instanceof Amount) {
    value = new BigNumber(joinLoHi(input.lo, input.hi).toString());
  }
```

File: packages/types/src/pnum.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    toRoundedNumber(decimals = exponent): number {
+      return Number(round({ value: value.shiftedBy(-exponent).toNumber(), decimals }));
+    },
+
+    toLoHi(): LoHi {
+      return splitLoHi(BigInt(value.toString()));
+    },
+
+    toAmount(): Amount {
+      return new Amount(splitLoHi(BigInt(value.toString())));
```
Comment (grod220) [inline packages/types/src/pnum.ts:110]: issue: missing tests for this method

File: .changeset/modern-taxis-repair.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+'@penumbra-zone/types': minor
+---
```
Comment (grod220) [inline .changeset/modern-taxis-repair.md:3]: question: should we take those packages you've listed, create a new package, make those specific files private, and only expose pnum?

File: packages/types/src/pnum.ts [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+ * Likewise for all methods, the outputs are
+ * - in base units for bigint, LoHi, Amount and ValueView
+ * - in display form with decimals for string and number
+ *
+ * @param input
+ * @param exponent
+ */
+function pnum(
+  input: string | number | LoHi | bigint | Amount | ValueView | undefined,
+  exponentInput = 0,
```
Comment (grod220) [inline packages/types/src/pnum.ts:28]: issue: think we should consider an options obj as a second parameter so it's easier to expand in the future:

```diff
- exponentInput = 0,
+ options: { exponentInput?: number } = { exponentInput: 0 },
```

File: packages/types/src/pnum.ts [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+ *
+ * Likewise for all methods, the outputs are
+ * - in base units for bigint, LoHi, Amount and ValueView
+ * - in display form with decimals for string and number
+ *
+ * @param input
+ * @param exponent
+ */
+function pnum(
+  input: string | number | LoHi | bigint | Amount | ValueView | undefined,
```
Comment (grod220) [inline packages/types/src/pnum.ts:27]: question: what are the usecases for an `undefined` input?

File: .changeset/modern-taxis-repair.md (hunk 4 above)
Comment (grod220) [reply .changeset/modern-taxis-repair.md:3]: related: if we make the other stuff private, we'll need to find the previous usages and convert them to pnum

**Review by grod220 — COMMENTED** [review]
Cool stuff!

File: packages/types/src/pnum.ts (hunk 1 above)
Comment (JasonMHasperhoven) [reply packages/types/src/pnum.ts:29]: I did this on purpose as it's shorter: `pnum(val)` vs `new PNum(val)`. My guess is that we'll use it quite often and this way it looks cleaner imo, especially if we need to do it inline inside a react component.

File: packages/types/src/pnum.ts (hunk 2 above)
Comment (JasonMHasperhoven) [reply packages/types/src/pnum.ts:44]: This case catches an instance of Amount **and LoHi**, in your check a LoHi will not be true.

File: packages/types/src/pnum.ts (hunk 5 above)
Comment (JasonMHasperhoven) [reply packages/types/src/pnum.ts:28]: Perhaps wise indeed

File: packages/types/src/pnum.ts (hunk 6 above)
Comment (JasonMHasperhoven) [reply packages/types/src/pnum.ts:27]: This: 
image

I guess you can make an argument against it, what do you think?

File: packages/types/src/pnum.ts (hunk 2 above)
Comment (grod220) [reply packages/types/src/pnum.ts:44]: No test appears to fail when removing that line. Can you add a test for coverage for that?

**Review by grod220 — APPROVED** [review]
Feel free to merge after addressing the final comments 👍 

Think it's mostly more tests we need

**Comment by TalDerei** [comment]
nice helper function, I've been wanting something like this for a while 👍

File: packages/types/src/pnum.ts (hunk 2 above)
Comment (JasonMHasperhoven) [reply packages/types/src/pnum.ts:44]: Sure

File: packages/types/src/round.ts [hunk 7]
```diff
@@ -4,9 +4,10 @@ import { removeTrailingZeros } from './shortify.js';
 export type RoundingMode = 'half-up' | 'up' | 'down';
 
 export interface RoundOptions {
-  value: number;
+  value: number | string;
```
Comment (JasonMHasperhoven) [inline packages/types/src/round.ts:7]: For @grod220 (because I know you want to know why) this is needed for a big number (caught by one of the tests)

**Pull request merged**